cmd/atlas/internal/cmdapi/diff.go

```diff
@@ -17,85 +24,296 @@ import (
-			cmdDiffRun(cmd, &opts)
+		Long: `'atlas schema diff' reads the state of two given schema definitions, 
+calculates the difference in their schemas, and prints a plan of
+SQL statements to migrate the "from" database to the schema of the "to" database.
+The database states can be read from a connected database, an HCL project or a migration directory.`,
+		Example: `  atlas schema diff --from mysql://user:pass@localhost:3306/test --to file://schema.hcl
+  atlas schema diff --from mysql://user:pass@localhost:3306 --to file://schema_1.hcl --to file://schema_2.hcl
+  atlas schema diff --from mysql://user:pass@localhost:3306 --to file://migrations`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return cmdDiffRun(cmd, &opts)
 		},
 	}
 	cmd.Flags().StringVarP(&opts.fromURL, "from", "", "", "[driver://username:password@protocol(address)/dbname?param=value] select a database using the URL format")
 	cmd.Flags().StringVarP(&opts.toURL, "to", "", "", "[driver://username:password@protocol(address)/dbname?param=value] select a database using the URL format")
+	cmd.Flags().StringVarP(&opts.devURL, "dev-url", "", "", "[driver://username:password@protocol(address)/dbname?param=value] select a database using the URL format")
+	cmd.Flags().SortFlags = false
 	cobra.CheckErr(cmd.MarkFlagRequired("from"))
 	cobra.CheckErr(cmd.MarkFlagRequired("to"))
 	return cmd
 }
 
 func init() {
-	diffCmd := newDiffCmd()
-	schemaCmd.AddCommand(diffCmd)
+	schemaCmd.AddCommand(newDiffCmd())
 }
 
 // cmdDiffRun connects to the given databases, and prints an SQL plan to get from
 // the "from" schema to the "to" schema.
-func cmdDiffRun(cmd *cobra.Command, flags *diffCmdOpts) {
-	ctx := cmd.Context()
-	fromC, err := sqlclient.Open(cmd.Context(), flags.fromURL)
-	cobra.CheckErr(err)
-	defer fromC.Close()
-	toC, err := sqlclient.Open(cmd.Context(), flags.toURL)
-	cobra.CheckErr(err)
-	defer toC.Close()
-	fromS := fromC.URL.Schema
-	toS := toC.URL.Schema
+func cmdDiffRun(cmd *cobra.Command, flags *diffCmdOpts) error {
+	var (
+		ctx = cmd.Context()
+		c   *sqlclient.Client
+	)
+	// We need a driver for diffing and planning. If given, dev database has precedence.
+	if flags.devURL != "" {
+		var err error
+		c, err = sqlclient.Open(ctx, flags.devURL)
+		if err != nil {
+			return err
+		}
+		defer c.Close()
+	}
+	from, err := stateReader(ctx, &stateReaderConfig{urls: []string{flags.fromURL}, dev: c})
+	if err != nil {
+		return err
+	}
+	defer from.Close()
+	to, err := stateReader(ctx, &stateReaderConfig{urls: []string{flags.toURL}, dev: c})
+	if err != nil {
+		return err
+	}
+	defer to.Close()
+	if c == nil {
+		var ok bool
+		c, ok = from.Closer.(*sqlclient.Client)
+		if !ok {
+			c, ok = to.Closer.(*sqlclient.Client)
+			if !ok {
+				// There is no driver to diff. We need a dev-database.
+				cobra.CheckErr(errors.New(
+					"neither --from nor --to provides a database connection, please provide one by passing in a --dev-url",
+				))
```

Comment:
This check occurs 3 times in this code.

Also, we can't assume the database can be used as a dev connection because we create and destroy resources there. My suggestion is to ask for `--dev-url` in case a file is being used in one of the states. This will eliminate the multiple checks below.